### PR TITLE
[Frontend] Fixing pytests and local wheel on macOS

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Another patch for dealing with differences between OSes. 

**Description of the Change:**

1. The `cp` command default behaviour on BSD is sufficient. On Linux we need the flag `--dereference`. BSD's `cp` does not contain a `--dereference flag`.
2. The `ld64` linker will succeed even if it attempts to link something unknown. Add flag `-arch_errors_fatal` to the linker.
3. Add parenthesis in one of the tests. For some reason, there appears to be a difference in behaviour in one of the stages. Non-stochastically it would result in undefined wires. We need to look more into this.
4. Add `-lc++` to a test to guarantee symbols are present during linking.

### CODE REVIEW

File: frontend/test/pytest/test_compiler.py [hunk 1]
```diff
@@ -177,7 +178,9 @@ class CompileCXXException:
             """Class that overrides the program to be compiled."""
 
             _executable = "cc"
-            _default_flags = ["-shared", "-fPIC", "-x", "c++"]
+
+            libcpp = "" if platform.system() == "Linux" else "-lc++"
```
Comment (dime10) [inline frontend/test/pytest/test_compiler.py:182]: A comment for this flag might be nice

File: frontend/test/pytest/test_pytree_args.py [hunk 2]
```diff
@@ -481,7 +481,7 @@ def test_args_used_indirectly_in_measure(self, backend):
         def circuit(dictionary):
             """q0 = 1; q1 = 0;"""
             qml.RX(jnp.pi, wires=0)
-            return measure(dictionary["wire"] + 1 % 2)
+            return measure((dictionary["wire"] + 1) % 2)
```
Comment (dime10) [inline frontend/test/pytest/test_pytree_args.py:484]: According to [this](https://docs.python.org/3/reference/expressions.html#operator-precedence), `%` comes before `+`, so the added parenthesis actively change the intended behaviour. Taking `1 % 2` doesn't make much sense, so I'm guessing this was just a bug in the test previously?
What prompted the change?

Comment (erick-xanadu) [reply frontend/test/pytest/test_pytree_args.py:484]: Yes, this is an interesting change that I have yet to narrow down. On macOS the runtime would raise an exception saying that a measurement was performed on an invalid wire. This exception is correct according to the operator precedence you link to. It is a bit concerning that:

1. It succeeded in Linux
2. On macOS, without the parenthesis, I believe the failure / success was not deterministic.

I have seen similar behaviour in the past (other compiler passes I was working on) when there was some undefined behaviour and a non-deterministic iteration was performed over some structures in the compiler (all functions in the module for example). But that seems a bit unlikely here.

File: setup.py [hunk 3]
```diff
@@ -66,6 +66,7 @@
     ext._add_ldflags(["-L", lib_path_npymath])  # pylint: disable=protected-access
     ext._add_ldflags(["-lnpymath"])  # pylint: disable=protected-access
     ext._add_cflags(["-I", np.get_include()])  # pylint: disable=protected-access
+    ext._add_cflags(["-std=c++11"])  # pylint: disable=protected-access
```
Comment (maliasadi) [inline setup.py:69]: Why is this required?

**Review by maliasadi — APPROVED** [review]

File: setup.py (hunk 3 above)
Comment (erick-xanadu) [reply setup.py:69]: This is necessary on older versions of clang that do not compile with c++11 by default. The include files of the wrapper (which add the Numpy header files), will use `constexpr` which was added in c++11.

Comment (maliasadi) [reply setup.py:69]: Why not supporting a more up-to-date standard, c++17 or c++20?

File: frontend/test/pytest/test_pytree_args.py (hunk 2 above)
Comment (dime10) [reply frontend/test/pytest/test_pytree_args.py:484]: Just noticed that it's even worse, looking at the input we use (`0`) the two cases should be identical (`0 + (1 % 2)` and `(0 + 1) % 2`) 👀 

I think it would be good to investigate further in this case, since there shouldn't be a "qubit" out of bounds error.

File: frontend/test/pytest/test_compiler.py (hunk 1 above)
Comment (dime10) [reply frontend/test/pytest/test_compiler.py]: Nice thanks!

File: setup.py (hunk 3 above)
Comment (dime10) [reply setup.py:69]: Correct me if I'm wrong, but specifying the minimum version necessary for a piece of code to compile seems appropriate. If the code only uses c++-11 features, than there is no need to be more restrictive by requiring a compiler with c++-20 support.

File: frontend/test/pytest/test_pytree_args.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_pytree_args.py:484]: What do you mean the input we use is 0? Some times it uses 1.

Comment (erick-xanadu) [reply frontend/test/pytest/test_pytree_args.py:484]:
```python
        result = circuit({"wire": 0})
        assert jnp.allclose(result, False)
        result = circuit({"wire": 1})
        assert jnp.allclose(result, True)
 ```

File: setup.py (hunk 3 above)
Comment (erick-xanadu) [reply setup.py:69]: This is the bare minimum. [`clang` 16 switched to use c++17 by default](https://clang.llvm.org/cxx_status.html#:~:text=By%20default%2C%20Clang%2016%20or,in%20Clang%204%20and%20earlier). The `clang` binary shipped on macOS is `clang` 14. I just doubled checked, and it is pybind headers (included from Numpy headers) which use constexpr. [pybind11 is compatible with `c++11`.](https://pybind11.readthedocs.io/en/stable/compiling.html#configuration-variables) So I think we should keep `c++11`

Comment (maliasadi) [reply setup.py:69]: My point here is with consistency and c++11 is a pretty old standard. We should ensure the code builds with either C++17 or C++20 support. There isn't any  issue with updating this to C++17.
  ```suggestion
    ext._add_cflags(["-std=c++17"])  # pylint: disable=protected-access
```

Comment (maliasadi) [reply setup.py:69]: it is a small change and of course this's possible to compile separate targets with different standards, but it is best to be consistent. We agreed C++17 to be the minimum C++ standard and I think we can keep it as is across the project.

Comment (erick-xanadu) [reply setup.py:69]: Ok, we had agreed in the past about c++17, that's a good enough reason for me.

File: frontend/test/pytest/test_pytree_args.py (hunk 2 above)
Comment (dime10) [reply frontend/test/pytest/test_pytree_args.py:484]: Oh my bad, must have missed the last bit of code there.

Maybe it is a bug in the runtime? I can confirm that
```py
@qjit
@qml.qnode(qml.device("lightning.qubit", wires=2))
def circuit(x: int):
    return measure(x)
circuit()
```
works for both `0`, `1`, and `2`, but not `3`.

File: setup.py (hunk 3 above)
Comment (dime10) [reply setup.py]: Generally I agree with Erick. But if you insist on a consistent standard that's fine by me too. Note though that you already specify C++20 in the runtime, not 17.

File: frontend/test/pytest/test_pytree_args.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_pytree_args.py:484]: I'm not sure yet. I have confirmed that with parenthesis, the MLIR generated actually generates a `remainder` function. Without it, the `remainder` function gets optimized away before MLIR generation. So, the test is more correct after this PR. The question about why it succeeded when it shouldn't, is still open.

Comment (erick-xanadu) [reply frontend/test/pytest/test_pytree_args.py:484]: @dime10 I think this is a bug of its own, so I will just open an issue and resolve this conversation if there is nothing else.

Comment (dime10) [reply frontend/test/pytest/test_pytree_args.py:484]: The OS thing is really weird, I wonder if there is a difference in the IR generated between the two.

If the order of operations is respected, `remainder` should not be part of the IR since `1 % 2` is just regular Python code that would be transparent to tracing process, unless either argument is a tracer.

**Comment by maliasadi** [comment]
[sc-43392]
